python_modules/dagster/dagster/_core/execution/plan/execute_step.py

```diff
@@ -766,7 +766,7 @@ def _store_output(
 ) -> Iterator[DagsterEvent]:
     output_def = step_context.op_def.output_def_named(step_output_handle.output_name)
     output_manager = step_context.get_io_manager(step_output_handle)
-    output_context = step_context.get_output_context(step_output_handle)
+    output_context = step_context.get_output_context(step_output_handle, output.metadata)
```

Comment:
i kinda am leaning towards `normalize_metadata_with_output` to make things very very explicit. but i dont have strong opinions. worth checking with other foundations folks maybe?

Reply:
yeah @alangenfeld  and @smackesey  interested in your thoughts on this

Reply:
>Are there any concerns with modifying the step context with the finalized dictionary of metadata?

This feels wrong to me. Having context be this stateful object that holds metadata is a bit of a mistake to begin with IMO, and this would be made worse by mutating it after execution is complete with the contents of a return `Output`.

But I agree it's a little weird to patch just `output.metadata` into `get_output_context`. However it feels very natural to pass the output object itself in (you can just pull off the metadata inside `get_output_context`), so that's what I favor.

>a community pr that is trying to access output metadata in the hook context

Not very familiar with how hook context is constructed but I'd also recommended a similar approach to ^^ rather than mutating `StepExecutionContext` with content of returned `Output`.

Reply:
> Having context be this stateful object that holds metadata is a bit of a mistake to begin with IMO

heard

a little context on the hook thing: the issue for hook context is that it doesn't have access to the output when it's created. maybe that's the real thing to fix. the community PR is trying to mimic what we do for passing output values to hooks where we store the value on the step context when we know there is a hook that could be invoked (https://github.com/dagster-io/dagster/blob/cd80ceb4e99f683010af324348cf6d6728897675/python_modules/dagster/dagster/_core/execution/plan/execute_step.py#L573). this is the exact same mistake that you point out. maybe it's a bit better than always updating the metadata since it would only happen in the case when there is a hook

Reply:
@smackesey  do you have any other thoughts on this? I think i'm inclined to keep this PR as is and eat the less-than-ideal pattern for the hook context. My justification is that we're already using the step context to store outputs when we have to make a hook context, so also storing metadata isn't introducing a brand new pattern, and hooks aren't a widely used feature so this code path won't be used in most execution

Reply:
Sorry for the late response. Can we pass just the `Output` (as `output`) instead of `output_metadata` into `get_output_context`? It makes sense that the full `Output` should be available when constructing the output context. Otherwise this PR looks good to me.

I am going to continue the hook discussion in that PR.